connection.go

```diff
@@ -1094,13 +1128,31 @@ func (conn *Connection) send(req Request, streamId uint64) *Future {
 		default:
 		}
 	}
+
+	if atomic.LoadUint32(&(conn.shutdownInProgress)) == shutdownInProgress {
+		conn.cancelFuture(fut, fmt.Errorf("server shutdown in progress"))
+		return fut
+	}
```

Comment:
It seems like there is no data race in such a solution:
```suggestion
        conn.shutdownWg.Add(1)
	if atomic.LoadUint32(&(conn.shutdownInProgress)) == shutdownInProgress {
		conn.cancelFuture(fut, fmt.Errorf("server shutdown in progress")) // conn.shutdownWg.Done() inside
		return fut
	}
	// conn.shutdownWg.Done() inside all finish future methods.
```
But in it looks like extra field. You could try to combine this into a new logical structure somehow:

https://github.com/tarantool/go-tarantool/blob/265f96c9ff5f39e436e7645845fea4589762174d/connection.go#L150-L151

Reply:
> But in it looks like extra field. You could try to combine this into a new logical structure somehow:

I failed to grasp what extra field you are talking about.

Reply:
https://github.com/tarantool/go-tarantool/blob/master/connection.go#L1163-L1167

Reply:
Moved decrement to `markDone`